src/transformStateWithClones.js

```diff
@@ -7,7 +7,63 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const arrayResults = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        if (arrayResults.length === 0) {
+          arrayResults.push(Object.assign({}, state, action.extraData));
+        } else {
+          arrayResults.push(
+            Object.assign(
+              {},
+              arrayResults[arrayResults.length - 1],
+              action.extraData,
+            ),
+          );
+        }
+        break;
+
+      case 'removeProperties':
+        if (arrayResults.length === 0) {
+          const newState = { ...state };
+
+          for (const key of action.keysToRemove) {
+            delete newState[key];
+          }
+          arrayResults.push(newState);
+        } else {
+          const newState = { ...arrayResults[arrayResults.length - 1] };
+
+          for (const key of action.keysToRemove) {
+            delete newState[key];
+          }
+          arrayResults.push(newState);
+        }
+        break;
+
+      case 'clear':
+        if (arrayResults.length === 0) {
+          const newStates = { ...state };
```

Comment:
The variable `newStates` is declared here, but it would be more consistent to use `newState` as in the other cases ('removeProperties' and 'addProperties'). This would improve code readability and consistency.